### PR TITLE
Implement x/oracle for JUNO

### PR DESCRIPTION
**Pull request by vuong177**
- Implement x/oracle for JUNO

### CODE REVIEW

File: price-feeder/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# Oracle Price Feeder
+
+The `price-feeder` tool is an extension of Juno's `x/oracle` module, both of
+which are based on Terra's [x/oracle](https://github.com/terra-money/classic-core/tree/main/x/oracle) and Umee [x/oracle](https://github.com/umee-network/tree/umee/x/oracle) module.
+core differences are as follows:
+
+- All exchange rates must be quoted in USD or USD stablecoins.
+- No need or use of reference exchange rates (e.g. Luna).
+- No need or use of Tobin tax.
+- The `price-feeder` combines both `feeder` and `price-server` into a single
```
Comment (adamewozniak) [inline price-feeder/README.md:10]: All of these are differences from the terra fork to the umee fork, everything else here looks the same

Comment (vuong177) [reply price-feeder/README.md:10]: We'll add Juno swap as price provider. Yep it's implemented from Umee.

File: x/oracle/abci.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+				return err
+			}
+		}
+
+		// update miss counting & slashing
+		voteTargetsLen := len(voteTargets)
+		claimSlice := types.ClaimMapToSlice(validatorClaimMap)
+		for _, claim := range claimSlice {
+			// Skip valid voters
+			if int(claim.WinCount) == voteTargetsLen {
```
Comment (codehans) [inline x/oracle/abci.go:66]: this line needs to be >=, otherwise if you get >67% VP voting for a non-whitelisted asset then everyone gets a miss

**Review by faddat — DISMISSED** [review]

File: price-feeder/README.md (hunk 1 above)
Comment (adamewozniak) [reply price-feeder/README.md:13]: ty for clarifying in docs @vuong177 💯

**Comment by faddat** [comment]
The gov prop for the oracle has passed:

https://www.mintscan.io/juno/proposals/57

We should look into getting this merged for v12.

**Comment by faddat** [comment]
Going through the error logs on the latest commit, I notice that we are having issues subscribing to the binance websocket. 

Kinda :red_flag: to me.

**Review by faddat — DISMISSED** [review]

**Review by faddat — APPROVED** [review]
lgtm

**Review by faddat — DISMISSED** [review]

**Review by faddat — DISMISSED** [review]

**Review by faddat — DISMISSED** [review]

**Review by faddat — DISMISSED** [review]

**Comment by faddat** [comment]
@vuong177 -- I'd like to get this merged, so that further PR's target the main branch.  This branch is getting huge.

**Review by faddat — APPROVED** [review]

**Comment by Reecepbcups** [comment]
I find it hard to stop working... anyways an easier way to review is probably also to compare base branch umee to the changes made here in juno, so I made this repo to do just that. 
Ill review this after the holiday & approve provided no major issues are found.

https://github.com/Reecepbcups/compare-oracle-changes/pull/1/files

**Comment by faddat** [comment]
@Reecepbcups nice review method.

**Review by faddat — APPROVED** [review]

File: x/oracle/types/asset.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	appparams "github.com/CosmosContracts/juno/v12/app/params"
+)
+
+const (
+	JunoDenom       string = appparams.BondDenom
+	JunoSymbol      string = "JUNO"
+	JunoExponent           = uint32(6)
+	AtomDenom       string = "ibc/6842C591DC4588411A565C9FF650FB15A17DFE3F0A43201E8141E4D14B8D171A"
```
Comment (Reecepbcups) [inline x/oracle/types/asset.go:13]: This is dig channel-37 not atom

```
junod q ibc-transfer denom-trace ibc/6842C591DC4588411A565C9FF650FB15A17DFE3F0A43201E8141E4D14B8D171A
denom_trace:
  base_denom: udig
  path: transfer/channel-37
```

ATOM channel-1 is `ibc/C4CFF46FD6DE35CA4CF4CE031E643C8FDC9BA4B99AE598E9B0ED98FE3A2319F9` for juno.

File: x/oracle/keeper/history_test.go [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+	require.Error(t, err)
+	require.Equal(t, types.PriceHistoryEntry{}, phStore)
+
+	phStores, err := oracleKeeper.getHistoryEntryBetweenTime(
+		ctx,
+		"Juno",
+		phEntrys[0].PriceUpdateTime.Add(-time.Minute),
+		phEntrys[2].PriceUpdateTime.Add(time.Minute),
+	)
+	require.NoError(t, err) // To discuss: should this throw error?
```
Comment (Reecepbcups) [inline x/oracle/keeper/history_test.go:577]: Doesn't pass test due to `no values in range`. not sure why this is not coming up on the CI test

**Review by Reecepbcups — CHANGES_REQUESTED** [review]

File: price-feeder/oracle/client/tx.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+	txf = txf.WithGas(adjusted)
+
+	unsignedTx, err := txf.BuildUnsignedTx(msgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	feeAmount, err := calculateFeeAmount(gasPrice, adjusted)
+	unsignedTx.SetFeeAmount(feeAmount)
```
[the remark this answers is not in the document]
Comment (Reecepbcups) [reply price-feeder/oracle/client/tx.go:38]: update - we bypass globalfee now

File: proto/juno/oracle/v1/query.proto [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  // ActiveExchangeRates returns all active denoms
+  rpc ActiveExchangeRates(QueryActiveExchangeRates)
+      returns (QueryActiveExchangeRatesResponse) {
+    option (google.api.http).get =
+        "/juno/oracle/v1/denoms/active_exchange_rates";
+  }
+
+  // PriceHistoryAtTime return price of denom at specific time
+  rpc PriceHistoryAtTime(QueryPriceHistoryAtTime) 
+      returns (QueryPriceHistoryAtTimeRespone) {
```
[the remark this answers is not in the document]
Comment (faddat) [reply proto/juno/oracle/v1/query.proto:34]: "regen files?"

[the remark this answers is not in the document]
Comment (faddat) [reply proto/juno/oracle/v1/query.proto]: quality finding

File: Makefile [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 BRANCH := $(shell git rev-parse --abbrev-ref HEAD)
 COMMIT := $(shell git log -1 --format='%H')
+BINDIR ?= $(GOPATH)/bin
+SIMAPP = ./app
```
[the remark this answers is not in the document]
Comment (faddat) [reply Makefile:6]: pretty sure it's simapp for reasons.  Could be wrong though.

@vuong177 ?

File: app/decorators/min_commission.go [hunk 8]
```diff
@@ -8,6 +8,12 @@ import (
 	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
 )
 
+var DefaultIsAppSimulation = false
+
+func UpdateSimsFlag(flag bool) {
```
[the remark this answers is not in the document]
Comment (faddat) [reply app/decorators/min_commission.go:13]: Not really sure what you mean but if you'd like to do it in an alternative style, go for it I figure.

This just seems to be declaring a bool to use and... allowing it to be used from decorators.DefaultIsAppSimulation?

File: x/oracle/keeper/history_test.go (hunk 4 above)
Comment (faddat) [reply x/oracle/keeper/history_test.go:577]: the test is an error check?

File: x/oracle/keeper/proposal_handler.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+	return nil
+}
+
+func isSubSet(super, sub types.DenomList) (unSubList types.DenomList, subList types.DenomList, isSubSet bool) {
+	if len(sub) == 0 {
+		return unSubList, subList, true
+	}
+
+	var matches int
```
[the remark this answers is not in the document]
Comment (faddat) [reply x/oracle/keeper/proposal_handler.go:135]: hi @vuong177 what do you think?

File: proto/juno/oracle/v1/query.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (Reecepbcups) [reply proto/juno/oracle/v1/query.proto]: Meant: make proto-all 
So it remakes auto gen file

File: Makefile (hunk 7 above)
[the remark this answers is not in the document]
Comment (Reecepbcups) [reply Makefile:6]: It’s simapp for a Simulation app, but should be APP for Juno, as this is not cosmos-sdk similation

File: x/oracle/keeper/history_test.go (hunk 4 above)
Comment (Reecepbcups) [reply x/oracle/keeper/history_test.go:577]: Yes, fails with "go test ./..." right now locally for me

File: app/keepers/keepers.go [hunk 10]
```diff
@@ -333,16 +349,19 @@ func NewAppKeepers(
...
+	wasmOpts = append(tfbindings.RegisterCustomPlugins(&appKeepers.BankKeeper, &appKeepers.TokenFactoryKeeper), wasmOpts...)
 
 	// Stargate Queries
 	accepted := wasmkeeper.AcceptedStargateQueries{
 		"/ibc.core.client.v1.Query/ClientState":    &ibcclienttypes.QueryClientStateResponse{},
 		"/ibc.core.client.v1.Query/ConsensusState": &ibcclienttypes.QueryConsensusStateResponse{},
 		"/ibc.core.connection.v1.Query/Connection": &ibcconnectiontypes.QueryConnectionResponse{},
+
+		// oracle query
+		"/juno.oracle.v1.Query/ExchangeRates": &oracletypes.QueryExchangeRatesResponse{},
```
[the remark this answers is not in the document]
Comment (faddat) [reply app/keepers/keepers.go:363]: The typo is the misspelling you mentioned?

File: proto/juno/oracle/v1/query.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (faddat) [reply proto/juno/oracle/v1/query.proto]: Ahhhh ok not like Regen networks.

Ok cool.  Resolved.

Btw.... Make proto all doesn't work on Mac again.

So this is:

Make proto all (till it fails)
Make proto gen (this works)

File: Makefile (hunk 7 above)
[the remark this answers is not in the document]
Comment (faddat) [reply Makefile:6]: Hang on lemme poke gaia

[the remark this answers is not in the document]
Comment (faddat) [reply Makefile:6]: A winner is you

File: app/keepers/keepers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Reecepbcups) [reply app/keepers/keepers.go:363]: yes

File: x/oracle/client/cli/tx.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+	)
+
+	return cmd
+}
+
+// GetCmdDelegateFeedConsent creates a Cobra command to generate or
+// broadcast a transaction with a MsgDelegateFeedConsent message.
+func GetCmdDelegateFeedConsent() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "delegate-feed-consent [operator] [feeder]",
```
Comment (Reecepbcups) [inline x/oracle/client/cli/tx.go:38]: Bit confusing, took me a few tries. 

Should change operator (which seems like to wants your valopper address) and [feeder] to [new_feeder_address]. The docs will specify but not entirely clear here (as someone who has never run an oracle before)

**Review by Reecepbcups — COMMENTED** [review]
Will also add testing via smart contract query panics on query, but instantiate is fine

```bash
junod q wasm contract-state smart $CONTRACT_ADDR '{"exchange_rate":{"denom":"juno"}}'
#where contract address is the oracle_querier.wasm in scripts/

Error: rpc error: code = InvalidArgument desc = Error calling the VM: Error executing Wasm: Wasmer runtime error: RuntimeError: Aborted: panicked at 'called `Result::unwrap()` on an `Err` value: UnsupportedRequest { kind: "'/cosmos.bank.v1beta1.Query/AllBalances' path is not allowed from the contract" }', /home/vuong/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-std-1.1.5/src/results/system_result.rs:50:9: query wasm contract failed: invalid request
```

File: x/oracle/types/params.go [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+	return Params{
+		VotePeriod:               DefaultVotePeriod,
+		VoteThreshold:            DefaultVoteThreshold,
+		RewardBand:               DefaultRewardBand,
+		RewardDistributionWindow: DefaultRewardDistributionWindow,
+		AcceptList:               DefaultAcceptList,
+		SlashFraction:            DefaultSlashFraction,
+		SlashWindow:              DefaultSlashWindow,
+		MinValidPerWindow:        DefaultMinValidPerWindow,
+		PriceTrackingList:        DefaultAcceptList,
```
[the remark this answers is not in the document]
Comment (faddat) [reply x/oracle/types/params.go:67]: Don't know

File: app/decorators/min_commission.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (faddat) [reply app/decorators/min_commission.go:13]: Bump?

[the remark this answers is not in the document]
Comment (Reecepbcups) [reply app/decorators/min_commission.go:13]: resolved in latest commit

File: x/oracle/types/params.go [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+	DefaultVotePeriod               = BlocksPerMinute / 2 // 30 seconds
+	DefaultSlashWindow              = BlocksPerWeek       // window for a week
+	DefaultRewardDistributionWindow = BlocksPerYear       // window for a year
+)
+
+// Default parameter values
+var (
+	DefaultVoteThreshold = sdk.NewDecWithPrec(50, 2) // 50%
+	DefaultRewardBand    = sdk.NewDecWithPrec(2, 2)  // 2% (-1, 1)
+	DefaultAcceptList    = DenomList{
```
[the remark this answers is not in the document]
Comment (vuong177) [reply x/oracle/types/params.go:37]: this is default, we can handler this by upgrade handler.

File: x/oracle/client/cli/tx.go (hunk 11 above)
Comment (Reecepbcups) [reply x/oracle/client/cli/tx.go]: Resolved in 9413c05585d6f9cd84b3d685f599a882c3d3ec11

File: x/oracle/client/cli/query.go [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+func GetCmdQueryTwapPrice() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "twap [denom] [startTime] [endTime]",
+		Short: "Query twap between period time",
+		Long: strings.TrimSpace(
+			`Query twap for pool. Start time must be unix time. End time can be unix time or duration.
+Example:
+$ junod q twap JUNO 2022-12-25T19:42:07.100Z 2022-12-25T20:42:07.100Z
```
[the remark this answers is not in the document]
Comment (Reecepbcups) [reply x/oracle/client/cli/query.go:377]: Fixed description in another commit

File: price-feeder/README.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
+
+
+# Oracle Price Feeder
+This is Juno version of [Umee's price feeder](https://github.com/umee-network/umee/tree/main/price-feeder), work with SDK 0.45.9.
```
Comment (JakeHartnell) [inline price-feeder/README.md:4]:
```suggestion
This is Juno's version of [Umee's price feeder](https://github.com/umee-network/umee/tree/main/price-feeder), intended to work with SDK 0.45.11.
```

**Comment by JakeHartnell** [comment]
Do we have any info about this binary `scripts/oracle/oracle_querier.wasm`? Would be nice to have a link in the README about it: https://github.com/CosmosContracts/juno/pull/329/files#diff-f22554e55833f81001f9eab94c6d57144e48d0d3f8496c10c3199a75db5a83bc

Like with a link to the source code.

File: x/oracle/keeper/history.go [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+	"fmt"
+	"time"
+
+	"github.com/CosmosContracts/juno/v12/x/oracle/types"
+	"github.com/CosmosContracts/juno/v12/x/oracle/util"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// storeHistorical data writes to the store, in all needed indexing.
+// TODO : testing
```
Comment (JakeHartnell) [inline x/oracle/keeper/history.go:14]: Remove TODO? I see some tests for this function.
```suggestion
```

**Review by JakeHartnell — APPROVED** [review]
Tremendous amount of work. Made a few minor comments, and agree with @Reecepbcups's asks.

Feel free to tackle those is follow up tickets or issues? Would love to get this in.

Tremendous amount of work! Thank you! Looking forward to battle testing this in testnet.

**Comment by Reecepbcups** [comment]
I will move my above ask into its own issue so we can open another PR for: **oracle upgrade handler & tweaks**

**Review by Reecepbcups — APPROVED** [review]

**Review by JakeHartnell — APPROVED** [review]

**Pull request merged**